### PR TITLE
Fallback-aware failure handling for data-bridge

### PR DESCRIPTION
**Pull request by scypio**
Implementation of new knotx feature (fragment processing faiilure handling) within data bridge

## Description
New failure handling feature was implemented in  https://github.com/Cognifide/knotx/issues/466
This PR allows data bridge to take advantage of the new approach and handle data bridge failures in a more robust way. 

## Motivation and Context
See core feature documentation

## Upgrade notes (if appropriate)
Requires https://github.com/Cognifide/knotx/issues/466
Integration tests to be added to knotx-stack project

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [X] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: core/src/main/java/io/knotx/databridge/core/impl/FragmentProcessor.java [hunk 1]
```diff
@@ -49,9 +52,23 @@ public FragmentProcessor(Vertx vertx, DataBridgeKnotOptions options) {
...
+                .map(serviceEntry::getResultWithNamespaceAsKey)
+                .doOnError(e -> storeErrorInFragment(fragmentContext.fragment(), e, serviceEntry.getName()))
+        )
         .reduce(new JsonObject(), JsonObject::mergeIn)
-        .map(results -> applyData(fragmentContext, results));
+        .map(results -> applyData(fragmentContext, results))
+        .onErrorReturn(e -> {
+          LOGGER.error("Fragment processing failed. Cause:{}\nRequest:\n{}\nFragmentContext:\n{}\n", e.getMessage(), request.getClientRequest(), fragmentContext);
+          fragmentContext.fragment().failure(DataBridgeKnotProxy.SUPPORTED_FRAGMENT_ID, e);
+          if (!fragmentContext.fragment().fallback().isPresent()) {
```
Comment (tomaszmichalak) [inline core/src/main/java/io/knotx/databridge/core/impl/FragmentProcessor.java]: I think it should not happen here. Data Bridge should set error on fragment. Then at the end Assembler should check fallbacks and do some logic if the fallback is not defined.

Comment (tomaszmichalak) [reply core/src/main/java/io/knotx/databridge/core/impl/FragmentProcessor.java]: Data Bridge is not responsible for template evaluation, it adds the JSON object from the data source to Knot.x Context.

File: core/src/main/java/io/knotx/databridge/core/impl/FragmentProcessor.java [hunk 2]
```diff
@@ -49,9 +52,23 @@ public FragmentProcessor(Vertx vertx, DataBridgeKnotOptions options) {
         .doOnNext(this::traceService)
         .flatMap(serviceEntry ->
             fetchServiceData(serviceEntry, request).toObservable()
-                .map(serviceEntry::getResultWithNamespaceAsKey))
+                .map(serviceEntry::getResultWithNamespaceAsKey)
+                .doOnError(e -> storeErrorInFragment(fragmentContext.fragment(), e, serviceEntry.getName()))
```
Comment (tomaszmichalak) [inline core/src/main/java/io/knotx/databridge/core/impl/FragmentProcessor.java]: Like it!

File: core/src/main/java/io/knotx/databridge/core/impl/FragmentProcessor.java (hunk 1 above)
Comment (scypio) [reply core/src/main/java/io/knotx/databridge/core/impl/FragmentProcessor.java]: The main idea behind this PR is to have a different way of processing errors for fragments, depending on Fragment configuration. 

No fallback configured -> error handling happens on Knot level. 
Fallback configured -> mark fragment as failed, error handling on Assembler level.

I added a dedicated `handleError` methot to `FragmentProcessor` to make this idea more comprehensible.

**Pull request merged**